src/main/java/io/vertx/core/http/impl/WebSocketImplBase.java

```diff
@@ -195,73 +219,116 @@ void handleFrame(WebSocketFrameInternal frame) {
       if (frameHandler != null) {
         frameHandler.handle(frame);
       }
+    }
+  }
 
-      if (textMessageHandler != null && (frame.type() == FrameType.TEXT
-        || (processingPartialTextMessage && frame.type() == FrameType.CONTINUATION))) {
-        handleTextFrame(frame);
+  private class FrameAggregator implements Handler<WebSocketFrameInternal> {
+
+    private Handler<String> textMessageHandler;
+    private Handler<Buffer> binaryMessageHandler;
+    private Buffer textMessageBuffer;
+    private Buffer binaryMessageBuffer;
+
+    @Override
+    public void handle(WebSocketFrameInternal frame) {
+      switch (frame.type()) {
+        case TEXT:
+          handleTextFrame(frame);
+          break;
+        case BINARY:
+          handleBinaryFrame(frame);
+          break;
+        case CONTINUATION:
+          if (textMessageBuffer != null && textMessageBuffer.length() > 0) {
+            handleTextFrame(frame);
+          } else if (binaryMessageBuffer != null && binaryMessageBuffer.length() > 0) {
+            handleBinaryFrame(frame);
+          }
+          break;
       }
+    }
 
-      if (binaryMessageHandler != null && (frame.type() == FrameType.BINARY
-        || (processingPartialBinaryMessage && frame.type() == FrameType.CONTINUATION))) {
-        handleBinaryFrame(frame);
+    private void handleTextFrame(WebSocketFrameInternal frame) {
+      Buffer frameBuffer = Buffer.buffer(frame.getBinaryData());
+      if (textMessageBuffer == null) {
+        textMessageBuffer = frameBuffer;
+      } else {
+        textMessageBuffer.appendBuffer(frameBuffer);
+      }
+      if (textMessageBuffer.length() > maxWebSocketMessageSize) {
+        int len = textMessageBuffer.length() - frameBuffer.length();
+        textMessageBuffer = null;
+        String msg = "Cannot process text frame of size " + frameBuffer.length() + ", it would cause message buffer (size " +
+            len + ") to overflow max message size of " + maxWebSocketMessageSize;
+        handleException(new IllegalStateException(msg));
```

Comment:
I think we should reset the buffer in the error case here.  If we don't, the internal buffer will keep growing forever if they continue to pass in continuation frames.  I thought I had added a test to WebsocketTest that would now fail (because we wouldn't be able to recover from a message that was too large)